### PR TITLE
Add systemverilog plugin to yosys

### PR DESCRIPTION
**Pull request by matthuszagh**
###### Description of changes

Add UHDM and Surelog packages, and then add systemverilog plugin to yosys.

To test the systemverilog plugin, I ran

```
yosys> plugin -i systemverilog
yosys> read_systemverilog top.sv
```

top.sv:
```verilog
module top #(
  parameter integer N = 2
) (
  input logic clk,

  input logic [3:0] a [0:N-1],
  input logic [3:0] b [0:N-1],
  output logic [3:0] c [0:N-1]
);
    genvar i;
    generate
        for (i = 0; i < N; i = i + 1) begin
            always_ff @(posedge clk) c[i] <= a[i] + b[i];
        end
    endgenerate
endmodule
```

The `design_introspection` plugin doesn't compile along with `sdc` and `xdc`, which depend on it. The rest do.

I also wasn't able to get the tests working for Surelog.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by hzeller** [comment]
Oh, very nice, I was not aware that someone already was working on this (so I sent #185626).

I am happy to abandon my PR if we can get this moved forward instead. In particular as you already added the systemverilog module to yosys, which I was about to do next.

File: pkgs/applications/science/logic/surelog/default.nix
Comment (hzeller) [inline pkgs/applications/science/logic/surelog/default.nix:36]: Running the unit tests is probably good. The default testing fires a whole regression test which is a bit much, but the following will just run the Unittests that run in a few seconds

```nix
  doCheck = true;
  checkPhase = ''
    runHook preCheck
    make -j $NIX_BUILD_CORES UnitTests
    ctest --output-on-failure
    runHook postCheck
    '';
```

File: pkgs/applications/science/logic/surelog/default.nix
Comment (hzeller) [inline pkgs/applications/science/logic/surelog/default.nix:25]: If you add `gperftools`, it will use tcmalloc which is a performance win.

Comment (hzeller) [reply pkgs/applications/science/logic/surelog/default.nix:25]: Ah, and that should of course go to the `buildInputs`, not `nativeBuildInputs`

File: pkgs/applications/science/logic/surelog/default.nix
Comment (hzeller) [inline pkgs/applications/science/logic/surelog/default.nix:30]: the jdk and python packages need to be in the nativeBuildInputs, as they are only needed at build time.

File: pkgs/applications/science/logic/surelog/default.nix
Comment (hzeller) [inline pkgs/applications/science/logic/surelog/default.nix:38]: there is also a pkg/ directory, which we should probably leave in lib/surelog (it is some pre-generated parsed UVM code).

The libflatbuffers and antlr runtime library it also creates is a little annoying, I'll try to upstream a change in which these are taken from the system instead of if available.

**Comment by matthuszagh** [comment]
Thanks for the review @hzeller! I think I got everything, but let me know if not.

File: pkgs/applications/science/logic/surelog/default.nix
Comment (hzeller) [inline pkgs/applications/science/logic/surelog/default.nix:50]: mmh, I mean that we should _keep_ the pkg, in lib/surelog, because I think surelog is expecting it there (though I am not entirely sure if it will find it on Nixos). It is its private set of pre-compiled bulit-ins that it needs to access in certain contexts. I'd probably keep it there for now and we can worry later if that needs to be adapted.

(Also, do we need to move the libraries from `lib/surelog` to `lib` ? I suspect that is so that we can uniformly use it in Nix, so more a question of understanding than critique.
If we do, do we get into trouble for surelog also copying the libflatbuffers and antlr runtime it creates ? I'd probably be more comfortable if they were all together in one private surelog namespace, but I suppose Nix provides us that namespace by definition.
(sorty, Nix noob here))

File: pkgs/applications/science/logic/surelog/default.nix
Comment (hzeller) [inline pkgs/applications/science/logic/surelog/default.nix:58]: We should exclude darwin for now, it seems to be crashing on that platform, see the 'neutral' test.
( https://github.com/NixOS/nixpkgs/pull/173225/checks?check_run_id=7942907230 )

File: pkgs/applications/science/logic/surelog/default.nix
Comment (hzeller) [inline pkgs/applications/science/logic/surelog/default.nix:14]: For Version number: Surelog actually has a built-in version number
(https://github.com/chipsalliance/Surelog/blob/120a9d89001c8ecc2c63fdabee56538476c6f9f7/src/CommandLine/CommandLineParser.cpp#L66 ), which gets advanced every now and then.

..and that is what I am using in #185626 (which I'll abandon if we can get your change submitted).

Unfortunately, it is not tagged there in github yet with that version number, so currently we have to manually get it out of there, but I hope that tagging happens upstream soon.

So in order to avoid some weird version jump later (2022.xx.yy -> 1.36, which is a weird ordering as 2022 >> 1.), maybe we should now start using that version ? Meanwhile I attempt to get upstream to add the tags, so that it might make things simpler here in the future.

It might also not be problematic at all to have weird version changes, so we can also wait until that upstream tagging is happening.

Up to you.

**Review by hzeller — APPROVED** [review]
LGTM

Mostly comments that might or might not be important now (versioning and a location of an installed file), and that we certainly can fix later.

I'd disable the darwin platform right now, as it doesn't seem to build/run properly.

File: pkgs/applications/science/logic/surelog/default.nix
Comment (thoughtpolice) [reply pkgs/applications/science/logic/surelog/default.nix:14]: I don't have time today to do a more thorough review but, yes: if there are already some built-in tags that get advanced every now and then, let's use that, and try to encourage upstream to update more frequently as needed, rather than using random dates. In particular it makes the support/QA process much simpler to have aligned version numbers.

If we need to switch version schemes I don't think it's that terrible (it kind of messes with `nix-env`'s `compareVerions` path, but that's normally OK) but it would be nice to get it right the first time.

Comment (hzeller) [reply pkgs/applications/science/logic/surelog/default.nix:14]: I now got upstream to create a tag so the following should now work:

```
  pname = "surelog";
  version = "v1.36";

  src = fetchFromGitHub {
    owner = "chipsalliance";
    repo = "Surelog";
    rev = version;
    hash = "sha256-gMxpnSsJTXmg2VU7mQcw1J40OuqW6oo9/sGplLgV0ho=";
    fetchSubmodules = true;
  };
```

Comment (hzeller) [reply pkgs/applications/science/logic/surelog/default.nix:14]: w.r.t. the name: I prefer a canonical lowercase name for the derivation and just use the uppercase for the git repo name. But I leave that up to whatever style is preferred.

Comment (thoughtpolice) [reply pkgs/applications/science/logic/surelog/default.nix:14]: Yes, the name should be lowercased. Note that GitHub is actually case-insensitive when doing clones, so simply setting `pname = "surelog";` and then `fetchFromGitHub { repo = pname; }` will work fine and you don't need to duplicate the two, but honestly I don't think that duplication matters much, the repository name is extremely unlikely to change.

**Comment by hzeller** [comment]
Anything else you need to finalize this PR ?

File: pkgs/applications/science/logic/surelog/default.nix
Comment (matthuszagh) [reply pkgs/applications/science/logic/surelog/default.nix]: I _think_ I moved those libraries because they weren't being found otherwise, but I don't honestly remember. I can look into this a bit more closely, but it might take me a few days before I can get to it. In the meantime, I moved pkg back into lib/surelog.

**Comment by matthuszagh** [comment]
> Anything else you need to finalize this PR ?

Sorry for the delay @hzeller, and thanks for the review. Most issues should now be resolved.

File: pkgs/applications/science/logic/surelog/default.nix
Comment (hzeller) [inline pkgs/applications/science/logic/surelog/default.nix:14]: In the meantime, there is now a v1.37. Which is interesting, because now there is also a corresponding tag `v0.9.1.37` in UHDM, which was previously untagged.

File: pkgs/applications/science/logic/uhdm/default.nix
Comment (hzeller) [inline pkgs/applications/science/logic/uhdm/default.nix:10]: There is now a tag [v0.9.1.37](https://github.com/chipsalliance/UHDM/releases/tag/v0.9.1.37) in UHDM.

So we can have here a proper version
```nix
version = "0.9.1.37";
# and below
src = ... {
    ref = "v${version}";
};
```

File: pkgs/applications/science/logic/surelog/default.nix
Comment (hzeller) [inline pkgs/applications/science/logic/surelog/default.nix:14]: I learned in a review in one of my own packages by some nix maintainer that the preferred way in nix packages is to not include the 'v' here in the version name, but then assemble the full tag name below in the `rev` of the github repo.

```nix
version = "1.37"
# ... later
src = ... {
   ref = "v${version}";
   # ...
};
```

**Review by hzeller — COMMENTED** [review]
Good news, UHDM now also has a tag version, so we can use that and don't have to pull from a particular revision.

**Comment by hzeller** [comment]
Just checking how are things going :) I really would like to see Surelog in nix pkgs.

**Comment by matthuszagh** [comment]
Thanks for the ping @hzeller, made the changes, let me know.

File: pkgs/development/compilers/yosys/plugins/symbiflow.nix
Comment (hzeller) [inline pkgs/development/compilers/yosys/plugins/symbiflow.nix:22]: Maybe it is worthwhile updating this to latest head as well (I guess this PR has sitting for a while)

**Comment by hzeller** [comment]
This looks really good. I'll try today at home your branch to see everything building and working together.
If things work find, I'll see that I can get together a bunch of people that can approve.

**Comment by hzeller** [comment]
I got things compiled locally (patched in your changes to a local copy of nixpkgs and build)

Initially I got some issue with the symbiflow plugin, but I fixed it by getting to the latest (which also fixed the need to use `which`). So I suggest you apply the patch below.

I did not do functional tests yet with everything together, will do that tomorrow.

```patch
--- a/pkgs/development/compilers/yosys/plugins/symbiflow.nix
+++ b/pkgs/development/compilers/yosys/plugins/symbiflow.nix
@@ -4,7 +4,6 @@
 , python3
 , readline
 , stdenv
-, which
 , yosys
 , zlib
 , yosys-symbiflow
@@ -15,11 +14,11 @@
   src = fetchFromGitHub {
     owner  = "chipsalliance";
     repo   = "yosys-f4pga-plugins";
-    rev    = "aadd1735b2b7af0472e56dc23f1035d6e1904712";
-    hash   = "sha256-gmpEx+8XDN7f+6e/YG25VKgdadwqApPqU3S6OB4AryA=";
+    rev    = "27208ce08200a5e89e3bd4f466bc68824df38c32";
+    hash   = "sha256-S7txjzlIp+idWIfp/DDOznluA3aMFfosMUt5dvi+g44=";
   };
 
-  version = "2022.05.13";
+  version = "2022.09.27";
 
   # Supported symbiflow plugins.
   #
@@ -49,7 +48,7 @@ in lib.genAttrs plugins (plugin: stdenv.mkDerivation (rec {
   inherit src version plugin;
   enableParallelBuilding = true;
 
-  nativeBuildInputs = [ which python3 ];
+  nativeBuildInputs = [ python3 ];
   buildInputs = [ yosys readline zlib uhdm surelog ];
 
   # xdc has an incorrect path to a test which has yet to be patched
```

**Comment by hzeller** [comment]
@ofborg build uhdm surelog yosys-symbiflow

**Comment by hzeller** [comment]
The ofbuild fail report in https://logs.nix.ci/?key=nixos/nixpkgs.173225&attempt_id=5843089f-29c4-402c-9b2a-ea671596842a looks like it is the one that I observed can be fixed with the patch suggested in [the comment above](https://github.com/NixOS/nixpkgs/pull/173225#issuecomment-1259098627)

Investigating other ofborg issues now as it looks like they work fine on my machine.

File: pkgs/applications/science/logic/surelog/default.nix
Comment (hzeller) [inline pkgs/applications/science/logic/surelog/default.nix:20]: Updated 1.40
```suggestion
    hash = "sha256-5nhJilFIJJDCnJUEUgyPNtWSQUgWcvM6LDFgFatAl/k=";
```

File: pkgs/applications/science/logic/surelog/default.nix
Comment (hzeller) [inline pkgs/applications/science/logic/surelog/default.nix:14]:
```suggestion
  version = "1.40";
```

File: pkgs/applications/science/logic/uhdm/default.nix
Comment (hzeller) [inline pkgs/applications/science/logic/uhdm/default.nix:10]:
```suggestion
  version = "0.9.1.40";
```

File: pkgs/applications/science/logic/uhdm/default.nix
Comment (hzeller) [inline pkgs/applications/science/logic/uhdm/default.nix:16]:
```suggestion
    hash = "sha256-CliKU2WM8B9012aDcS/mTyIf+JcsVsc4uRRi9+FRWbM=";
```

**Review by hzeller — CHANGES_REQUESTED** [review]
I've upstreamed the change that should make the UHDM compile issue go away. Now in v0.9.1.40 version of UHDM.

Also, with the other patch in one of the review comments it should compile that.

**Comment by hzeller** [comment]
Confirmed working. After
```
nix-build -A yosys-symbiflow
```

Using yosys and directly importing the just generated Surelog systemverilog plugin, using the new `read_systemverilog` command works as expected:

```
$ yosys
# ...
 Yosys 0.20 (git sha1 4fcb95ed087, gcc 11.3.0 -fPIC -Os)

yosys> plugin -i /nix/store/1hcspmyzms91gls830chgm7079k9hxm3-yosys-symbiflow-systemverilog-plugin-2022.09.27/share/yosys/plugins/systemverilog.so

yosys> read_systemverilog "foo.sv"
1. Executing Verilog with UHDM frontend.
[INF:CM0023] Creating log file ./slpp_all/surelog.log.
[WRN:PA0205] foo.sv:1:1: No timescale set for "foo".

[INF:CP0300] Compilation...

[INF:CP0303] foo.sv:1:1: Compile module "work@foo".

[INF:CP0302] Compile class "work@mailbox".

[INF:CP0302] Compile class "work@process".

[INF:CP0302] Compile class "work@semaphore".

[INF:EL0526] Design Elaboration...

[NTE:EL0503] foo.sv:1:1: Top level module "work@foo".

[NTE:EL0508] Nb Top level modules: 1.

[NTE:EL0509] Max instance depth: 1.

[NTE:EL0510] Nb instances: 1.

[NTE:EL0511] Nb leaf instances: 1.

[INF:UH0706] Creating UHDM Model...

[  FATAL] : 0
[ SYNTAX] : 0
[  ERROR] : 0
[WARNING] : 1
[   NOTE] : 5
Generating RTLIL representation for module `\foo'.

yosys>
```

**Comment by matthuszagh** [comment]
All changes should be in.

File: pkgs/applications/science/logic/uhdm/default.nix
Comment (hzeller) [inline pkgs/applications/science/logic/uhdm/default.nix:19]: There is something strange [going on in darwin](https://github.com/NixOS/nixpkgs/pull/173225/checks?check_run_id=8585878593) where it fails to compile as capnp does not seem to find the biuld prefix and then does not produce the generated file. Maybe something different going on on that platform with symbolic links ?

I suggest that we attempt to add this and see if that makes compilation on Darwin happy.

```suggestion
  prePatch = ''                                                                                                                                                                                               
     substituteInPlace CMakeLists.txt --replace \                                                                                                                                                             
       'capnp compile' \                                                                                                                                                                                      
       'capnp compile --src-prefix=''${GENDIR}/..'                                                                                                                                                            
  '';

```

**Comment by hzeller** [comment]
@ofborg build yosys-symbiflow

**Review by hzeller — APPROVED** [review]
Thanks, I think this looks good. I am about to upstream the changes for the capnp compile patch upstream, so in future version this would not be needed anymore.

I have no idea yet why the aarrch tests are not fully working on darwin (I suspect we might need the `checkPhase = "TEST_TMPDIR=$TMPDIR make test";` in UHDM), but I leave that for later to iterate (I'll do some direct experiments so that we don't have to do this through a cumbersome review back and forth).

So, ready for submission!

**Review by ErinvanderVeen — APPROVED** [review]

**Comment by hzeller** [comment]
I am looking forward to have this submitted, because I am working on getting the upstream projects (surelog, uhdm) into a better state w.r.t. building and dependencies which I then like to reflect here with the next update.
This is a good base and good starting point, so we should get this submitted.

What needs to be done to get the attention of someone who can submit ?

**Review by aherrmann — APPROVED** [review]
@fricklerhandwerk No reason to hold this from my end. I trust @ErinvanderVeen 's review.
But, my check mark is gray, so I fear this will require your approval or someone else with commit bit.

**Pull request merged**
